modules/reindex/src/internalClusterTest/java/org/opensearch/index/codec/ReindexCodecIT.java

```diff
@@ -0,0 +1,179 @@
+
+import static java.util.stream.Collectors.toList;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_BLOCKS_METADATA;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_BLOCKS_READ;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_BLOCKS_WRITE;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_READ_ONLY;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_READ_ONLY_ALLOW_DELETE;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertNoFailures;
+
+public class ReindexCodecIT extends ReindexTestCase {
+
+    public void testReindexingMultipleCodecs() throws InterruptedException, ExecutionException {
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        Map<String, String> codecMap = Map.of(
+            "best_compression",
+            "BEST_COMPRESSION",
+            "zstd_no_dict",
+            "ZSTD_NO_DICT",
+            "zstd",
+            "ZSTD",
+            "default",
+            "BEST_SPEED"
+        );
+
+        for (Map.Entry<String, String> codec : codecMap.entrySet()) {
+            assertReindexingWithMultipleCodecs(codec.getKey(), codec.getValue(), codecMap);
+        }
+
+    }
+
+    private void assertReindexingWithMultipleCodecs(String destCodec, String destCodecMode, Map<String, String> codecMap)
+        throws ExecutionException, InterruptedException {
+
+        final String index = "test-index" + destCodec;
+        final String destIndex = "dest-index" + destCodec;
+
+        // creating source index
+        createIndex(
+            index,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put("index.codec", "default")
+                .build()
+        );
+        ensureGreen(index);
+
+        final int nbDocs = randomIntBetween(2, 5);
+
+        // indexing with all 4 codecs
+        for (Map.Entry<String, String> codec : codecMap.entrySet()) {
+            indexWithDifferentCodecs(index, codec.getKey(), codec.getValue(), nbDocs);
+        }
+
+        // creating destination index with destination codec
+        createIndex(
+            destIndex,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put("index.codec", destCodec)
+                .build()
+        );
+
+        BulkByScrollResponse bulkResponse = new ReindexRequestBuilder(client(), ReindexAction.INSTANCE).source(index)
+            .destination(destIndex)
+            .refresh(true)
+            .waitForActiveShards(ActiveShardCount.ONE)
+            .get();
+
+        assertEquals(4 * nbDocs, bulkResponse.getCreated());
+        assertEquals(4 * nbDocs, bulkResponse.getTotal());
```

Comment:
The number '4' can be derived from codecMap size.

Reply:
Done